## Investigasi sampel data titanic berikut dengan cara :
1. Cek secara head, tail, sample, info lalu observasi apa yang bisa anda peroleh ?
2. Lakukan Statistical Summary dengan mengekstrak informasi yang didapat dari observasi anda ?
3. Cek apakah ada duplikat dan bagaimana handlenya ?
4. Cek apakah ada missing value, berapa persentasenya jika ada, dan bagaimana cara handlenya ?

## Import Libraries

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

# Load Data

In [ ]:
# import data
df = pd.read_excel('titanic.xlsx')
data = df.copy()

# Preliminary

In [ ]:
data.head()

survived                                             name     sex      age
0         1                    Allen, Miss. Elisabeth Walton  female  29.0000
1         1                   Allison, Master. Hudson Trevor    male   0.9167
2         0                     Allison, Miss. Helen Loraine  female   2.0000
3         0             Allison, Mr. Hudson Joshua Creighton    male  30.0000
4         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0000

In [ ]:
data.tail()

survived                                     name     sex   age
495         1  Mallet, Mrs. Albert (Antoinette Magnin)  female  24.0
496         0        Mangiavacchi, Mr. Serafino Emilio    male   NaN
497         0               Matthews, Mr. William John    male  30.0
498         0                Maybery, Mr. Frank Hubert    male  40.0
499         0                McCrae, Mr. Arthur Gordon    male  32.0

In [ ]:
data.sample(9)

survived                                               name     sex   age
229         1  Penasco y Castellana, Mrs. Victor de Satode (M...  female  17.0
405         0                                Enander, Mr. Ingvar    male  21.0
113         1                         Fortune, Miss. Mabel Helen  female  23.0
296         1  Thayer, Mrs. John Borland (Marian Longstreth M...  female  39.0
373         0                      Clarke, Mr. Charles Valentine    male  29.0
441         1                   Herman, Mrs. Samuel (Jane Laver)  female  48.0
61          1  Cavendish, Mrs. Tyrell William (Julia Florence...  female  76.0
476         0              Lahtinen, Mrs. William (Anna Sylfven)  female  26.0
394         0                               Denbury, Mr. Herbert    male  25.0

Observations:

1. All columns, except `name`, 'sex' column, are numeric
2. `survived` column seems to contain two distinct values (0,1)
3. `sex` is apparently also two distinct values (female, male)
4. No obvious defect on the data (column name vs its entries), all looks good

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  500 non-null    int64  
 1   name      500 non-null    object 
 2   sex       500 non-null    object 
 3   age       451 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 15.8+ KB


Observation:
1. Dari info diatas didapat bahwa terdapat 500 data atau 500 baris, dengan memiliki 4 kolom, dengan kolom survived, name, sex, dan age.

2. Semua baris tidak ada data missing/kosong (kecuali data age), dengan perolehan 9 baris terdapat mising/kosong. dtypes age akan dikerjakan nanti

3. Dengan tipe data kolom name, dan sex adalah object atau kategorical, serta age merupakan data type float atau numerical.

4. All dtypes seems ok. given the coressponding column name

# Stastical Summary

In [ ]:
data.columns

Index(['survived', 'name', 'sex', 'age'], dtype='object')

In [ ]:
categoricals = ['name', 'sex']

numerical = ['survived','age']



In [ ]:
data[numerical].describe()

survived         age
count  499.000000  450.000000
mean     0.539078   35.877593
std      0.498971   14.758183
min      0.000000    0.666700
25%      0.000000   24.000000
50%      1.000000   35.000000
75%      1.000000   47.000000
max      1.000000   80.000000

description:
1. nilai min dan max pada kolom age memiliki rentang yg sangat jauh, berbeda dengan kolom survived. hal ini akan dicek kembali nanti
2. nilai mean, dan mediaan pada dtype age menunjukan symetrical distribution karena memiliki rentang nilai yang hampir mirip. berbeda dengan nilai survived yang menunjukan right skew distribution
3. kolom survived, adalah kolom binary hanya memiliki nilai 0-1 dan tidak perlu di cek simmetricity. only need to check balance level.

In [ ]:
data[categoricals].describe()

name   sex
count                         499   499
unique                        499     2
top     McCrae, Mr. Arthur Gordon  male
freq                            1   288

describtion:
1. column name memiliki 499 unique value yang berbeda-beda atau nama yang berbeda, hanya 1 orang yang memiliki nama sama dengan yang lain
2. column sex memiliki nilai unique 2 value yakni ''female dan male''. frequensi jawaban adalah male dengan frekuensi 288 male, dan sisanya adalah female, sebanyak 212.

In [ ]:
categoricals

['name', 'sex']

In [ ]:
for col in categoricals:
  print(f"value counts of {col} column")
  print(data[col].value_counts(),'\n')

value counts of name column
name
McCrae, Mr. Arthur Gordon                                                             1
Allen, Miss. Elisabeth Walton                                                         1
Allison, Master. Hudson Trevor                                                        1
Lehmann, Miss. Bertha                                                                 1
Laroche, Mrs. Joseph (Juliette Marie Louise Lafargue)                                 1
Laroche, Mr. Joseph Philippe Lemercier                                                1
Laroche, Miss. Simonne Marie Anne Andree                                              1
Laroche, Miss. Louise                                                                 1
Lamb, Mr. John Joseph                                                                 1
Lahtinen, Rev. William                                                                1
Lahtinen, Mrs. William (Anna Sylfven)                                                 1

In [ ]:
for col in data.columns:
  print(f"=== {col} ===")
  print(data[col].value_counts(),'\n')

=== survived ===
survived
1    269
0    230
Name: count, dtype: int64 

=== name ===
name
McCrae, Mr. Arthur Gordon                                                             1
Allen, Miss. Elisabeth Walton                                                         1
Allison, Master. Hudson Trevor                                                        1
Lehmann, Miss. Bertha                                                                 1
Laroche, Mrs. Joseph (Juliette Marie Louise Lafargue)                                 1
Laroche, Mr. Joseph Philippe Lemercier                                                1
Laroche, Miss. Simonne Marie Anne Andree                                              1
Laroche, Miss. Louise                                                                 1
Lamb, Mr. John Joseph                                                                 1
Lahtinen, Rev. William                                                                1
Lahtinen, Mrs. William (Anna S

In [ ]:
len(data.drop_duplicates())/len(data)

1.0

In [ ]:
list(data.columns)

['survived', 'name', 'sex', 'age']

In [ ]:
duplicates = data[data.duplicated(keep=False)]

In [ ]:
duplicates

survived                            name     sex   age
104         1  Eustis, Miss. Elizabeth Mussey  female  54.0
349         1  Eustis, Miss. Elizabeth Mussey  female  54.0

In [ ]:
duplicates.groupby(list(data.columns)).size()

survived  name                            sex     age 
1         Eustis, Miss. Elizabeth Mussey  female  54.0    2
dtype: int64

In [ ]:
duplicates.groupby(list(data.columns)).size().reset_index(name='jumlah duplikat')

survived                            name     sex   age  jumlah duplikat
0         1  Eustis, Miss. Elizabeth Mussey  female  54.0                2

In [ ]:
# hitung jumlah duplikat dengan data column, ukuran jumlah duplikat, kemudian mengganti index nama column
duplicate_counts = duplicates.groupby(list(data.columns)).size().reset_index(name='jumlah_duplikat')

# Mengurutkan data duplicate
sorted_duplicates = duplicate_counts.sort_values(by='jumlah_duplikat', ascending=False)
print("Baris duplikat yang sudah diurutkan berdasarkan jumlah kemunculannya:")
sorted_duplicates

Baris duplikat yang sudah diurutkan berdasarkan jumlah kemunculannya:


survived                            name     sex   age  jumlah_duplikat
0         1  Eustis, Miss. Elizabeth Mussey  female  54.0                2

In [ ]:
data = data.drop_duplicates()

In [ ]:
len(data.drop_duplicates())/len (data)

1.0

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  500 non-null    int64  
 1   name      500 non-null    object 
 2   sex       500 non-null    object 
 3   age       451 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 15.8+ KB


# Missing Value Handling

In [ ]:
data.isna().sum()

survived     0
name         0
sex          0
age         49
dtype: int64

In [ ]:
# len= menghitung jumlah baris
total_row = len (data)

In [ ]:
# Memunculkan total nilai data dalam kolom
for column in data.columns:
  print("====={column}=====")
  display(data[column].value_counts())
  print()

====={column}=====


survived
1    270
0    230
Name: count, dtype: int64


====={column}=====


name
Eustis, Miss. Elizabeth Mussey                                                        2
Becker, Miss. Ruth Elizabeth                                                          1
Becker, Miss. Marion Louise                                                           1
Becker, Master. Richard F                                                             1
Beauchamp, Mr. Henry James                                                            1
Beane, Mrs. Edward (Ethel Clarke)                                                     1
Beane, Mr. Edward                                                                     1
Bateman, Rev. Robert James                                                            1
Banfield, Mr. Frederick James                                                         1
Ball, Mrs. (Ada E Hall)                                                               1
Baimbrigge, Mr. Charles Robert                                                        1
Bailey, Mr. Percy Andrew                                                              1
Ashby, Mr. John                                                                       1
Angle, Mrs. William A (Florence "Mary" Agnes Hughes)                                  1
Becker, Mrs. Allen Oliver (Nellie E Baumgardner)                                      1
Andrew, Mr. Frank Thomas                                                              1
Andrew, Mr. Edgardo Samuel                                                            1
Aldworth, Mr. Charles Augustus                                                        1
Abelson, Mrs. Samuel (Hannah Wizosky)                                                 1
Abelson, Mr. Samuel                                                                   1
Young, Miss. Marie Grice                                                              1
Wright, Mr. George                                                                    1
Woolner, Mr. Hugh                                                                     1
Wilson, Miss. Helen Alice                                                             1
Williams-Lambert, Mr. Fletcher Fellows                                                1
Williams, Mr. Richard Norris II                                                       1
Williams, Mr. Charles Duane                                                           1
Willard, Miss. Constance                                                              1
Widener, Mrs. George Dunton (Eleanor Elkins)                                          1
Widener, Mr. Harry Elkins                                                             1
Angle, Mr. William A                                                                  1
Clarke, Mr. Charles Valentine                                                         1
Christy, Mrs. (Alice Frances)                                                         1
Christy, Miss. Julie Rachel                                                           1
Chapman, Mrs. John Henry (Sara Elizabeth Lawry)                                       1
Chapman, Mr. John Henry                                                               1
Chapman, Mr. Charles Henry                                                            1
Carter, Rev. Ernest Courtenay                                                         1
Carter, Mrs. Ernest Courtenay (Lilian Hughes)                                         1
Carbines, Mr. William                                                                 1
Campbell, Mr. William                                                                 1
Cameron, Miss. Clear Annie                                                            1
Caldwell, Mrs. Albert Francis (Sylvia Mae Harbaugh)                                   1
Caldwell, Mr. Albert Francis                                                          1
Caldwell, Master. Alden Gates                                                         1
Beesley, Mr. Lawrence              


====={column}=====


sex
male      288
female    212
Name: count, dtype: int64


====={column}=====


age
24.0000    23
30.0000    20
36.0000    19
18.0000    14
42.0000    14
45.0000    14
35.0000    14
22.0000    12
28.0000    12
23.0000    11
21.0000    11
31.0000    11
48.0000    11
25.0000    11
29.0000    10
54.0000    10
50.0000    10
34.0000    10
47.0000    10
39.0000     9
27.0000     9
33.0000     8
49.0000     8
38.0000     8
19.0000     8
40.0000     8
55.0000     8
32.0000     7
60.0000     7
17.0000     7
26.0000     7
44.0000     6
52.0000     6
37.0000     6
58.0000     6
64.0000     5
46.0000     5
57.0000     5
51.0000     5
16.0000     4
53.0000     4
56.0000     4
43.0000     4
62.0000     4
61.0000     4
1.0000      3
8.0000      3
63.0000     3
41.0000     3
15.0000     2
71.0000     2
20.0000     2
65.0000     2
4.0000      2
6.0000      2
2.0000      1
0.9167      1
14.0000     1
11.0000     1
76.0000     1
59.0000     1
80.0000     1
28.5000     1
32.5000     1
70.0000     1
45.5000     1
13.0000     1
67.0000     1
12.0000     1
0.8333      1
0.6667      1
7.0000      1
3.0000      1
Name: count, dtype: int64

conclution:
1. age memiliki range score 14-18 for the total score


In [ ]:
total_rows = len(data)
total_rows

500

In [ ]:
data.columns

Index(['survived', 'name', 'sex', 'age'], dtype='object')

In [ ]:
total_rows = len(data)

In [ ]:
# menghitung data mising dalam column
# menghitung presentase dengan elemen yang hilang dibagi total data dikali 100
# .2f digunakan untuk memunculkan persentase dengan 2 desimal
for column in data.columns:
    missing_count = data[column].isna().sum()
    missing_percentage = (missing_count / total_rows) * 100
    print(f"Column '{column}' Has {missing_count} missing values ({missing_percentage:.2f}%)") # .2f means 2 decimal

Column 'survived' Has 0 missing values (0.00%)
Column 'name' Has 0 missing values (0.00%)
Column 'sex' Has 0 missing values (0.00%)
Column 'age' Has 49 missing values (9.82%)


conclution:
1. missing value below 20%, so we handle numerically with median, and handle categorical with mode. but the caegorical has no missing value, namely name, and sex

In [ ]:
total_rows = len(data)
total_rows

500

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  500 non-null    int64  
 1   name      500 non-null    object 
 2   sex       500 non-null    object 
 3   age       451 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 15.8+ KB


In [ ]:
data['name'].dtype

dtype('O')

In [ ]:
data['sex'].dtype

dtype('O')

In [ ]:
data['name'].mode()[0]

'Eustis, Miss. Elizabeth Mussey'

In [ ]:
data['sex'].mode()[0]

'male'

In [ ]:
data['age'].median()

35.0

In [ ]:
for column in data.columns:
    if data[column].dtype == 'object':
        data[column].fillna(data[column].mode()[0], inplace=True)
    else:data[column].fillna(data[column].median(), inplace=True)

<ipython-input-36-c3bbe8196c51>:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  else:data[column].fillna(data[column].median(), inplace=True)
<ipython-input-36-c3bbe8196c51>:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=Tru

In [ ]:
data.isna().sum()

survived    0
name        0
sex         0
age         0
dtype: int64

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  500 non-null    int64  
 1   name      500 non-null    object 
 2   sex       500 non-null    object 
 3   age       500 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 15.8+ KB
